improve DBFS

tpdenk opened the pull request:
Improve the DBFS in general.
The current DBFS is just a draft with a lot of work to be done.
This PR implements at least a few parts of this remaining work.

tpdenk committed:
dbfs: implement dbfs validation

Upon loading a database, the file structure as well as the content
of the tables.info file will be validated.
If validation fails, the structure cannot be loaded as a (*DBFS).

Also changed the content of the tables.info file to now also hold the
count of tables in the database.

tpdenk committed:
improve DBFS and implement insert (again)

Insert now works, and a (skipped) test is in this commit, which inserts 100000 datasets (seems to work).

tsatke commented on internal/engine/page/page.go:
```diff
@@ -260,25 +260,30 @@ func (p *Page) storeRawCell(rawCell []byte) error {
+		pivotKey = c.Key
+	}
+
+	insertIndex := sort.Search(len(slots), func(i int) bool {
```
Searching for an insert index opposed to sorting all slots and rewriting all of them improved insert performance by about 45%.
For 100_000 inserts, the time dropped from 20sec to 11sec.
Maybe this can be improved even more in the future.

tsatke replied:
We don't search for an existing slot. The method `sort.Search` returns an index, at which an element must be inserted (for the slice to remain sorted).
Inserting the element correctly and then storing all new slots on disk is faster and better than obtaining all slots, adding the new one to the end, sorting the whole slice (which is what we did before) and then writing them back to disk.

tpdenk committed:
fix store cell slot in page and linter warnings

Messed up the store mechanism by writing the correctly sorted slots in reverse order.
Fixed this, so that they are written in original order.
Performance improvement of 45% persists, this did fortunately not change with the fix.

tpdenk committed:
go mod tidy

tpdenk committed:
remove long running, skipped test

The removed test is a duplicate of another test, but with far more iterations. It is skipped because it took about 12 seconds to complete. We don't need that test.

tpdenk committed:
remove unused constants

tsatke commented on internal/engine/dbfs/dbfs.go:
```diff
@@ -42,7 +43,9 @@ func CreateNew(fs afero.Fs) (*DBFS, error) {
 
 // Load loads a DBFS from the given file system.
 func Load(fs afero.Fs) (*DBFS, error) {
-	// FIXME: perform checks
```
you commented on a removed line. the added line says
`if err := Validate(fs); err != nil {`
Somehow you only got one of the two commits in this PR I think

tsatke commented on internal/engine/dbfs/dbfs.go:
```diff
@@ -12,7 +12,8 @@ import (
 )
 
-const defaultPerm os.FileMode = 0600
+const defaultDirPerm os.FileMode = 0744
```
Kind of. `0744` stands for `rwxr--r--`, which means, that only the owner can create files in the directory and all others can only read.
`0644` stands for `rw-r--r--`, which means, that only the owner can edit the file, but all can read it.

tsatke commented on internal/engine/dbfs/schema_file.go:
```diff
@@ -0,0 +1,122 @@
+	columns      []table.Col
+}
+
+type schemaYaml struct {
```
```suggestion
// schemaYaml is an intermediate structure used for encoding
// a SchemaFile into yaml.
type schemaYaml struct {
```

tsatke commented on internal/engine/dbfs/schema_file.go:
```diff
@@ -0,0 +1,122 @@
+	Columns      []columnYaml `yaml:"columns"`
+}
+
+type columnYaml struct {
```
```suggestion
// columnYaml is an intermediate structure used for encoding
// a table.Col into yaml.
type columnYaml struct {
```

tsatke commented on internal/engine/dbfs/validate_test.go:
```diff
@@ -0,0 +1,175 @@
+			func(fs afero.Fs) {
+				mkdir(fs, TablesDirectory)
+				write(fs, `
+count: 1
```
That's how Go's multiline strings work. If I indent them, I have tabs in my string.
In Go,
```go
s := `a
bc`
```
and
```go
s := `a
    bc`
```
are different things.

tsatke commented on internal/engine/engine_test.go:
```diff
@@ -53,15 +54,34 @@ func (suite *EngineSuite) EqualTables(expected, got table.Table) {
+	var gotRows []table.Row
+	var gotErr error
+
+	for {
```
Order independent comparison.
First for loop collects all expected rows in memory, second loop all actual rows.
after that, line 82-84, it is checked that all actual rows are expected.
This works, because in line 80 there is a check that the amount of expected and actual rows match.

SUMUKHA-PK replied:
Can this be added there?

tsatke replied:
Probably not worth, as this will change as soon as we require sorted comparison for some test cases, i.e. as soon as we support `ORDER BY`.

tsatke commented on internal/engine/table.go:
```diff
@@ -41,9 +41,10 @@ type Inserter interface {
-	dataProvider func() (*dbfs.PagedFile, error)
+	name string
+
+	openSchemaFile func() (*dbfs.SchemaFile, error)
```
No because I wanna be able to set it. I can't do that if it's a method on the struct.

tpdenk committed:
Apply suggestions from code review

SUMUKHA-PK commented:
Just one more thing and we good to merge!
